File: CPP/src/Graph.cpp
//
// Created by cheily on 03.05.2023.
//
#include "Graph.h"


Node * Graph::goal = nullptr;

bool Graph::
isGoal(const Node & n) {
    for (int y = 0; y < n.size_y; ++y) {
        for (int x = 0; x < n.size_x; ++x) {
            if (n.table[y][x] != goal->table[y][x]) return false;
        }
    }
    return true;
}

std::array< std::optional<Node>, 4 > Graph::
neighbors(Node n, std::string order) {
    std::array< std::optional<Node>, 4 > ret;
    for (int i = 0; i < order.length(); i++) {
        if (order[i] == 'L') ret[i] = L(n);
        else if (order[i] == 'R') ret[i] = R(n);
        else if (order[i] == 'U') ret[i] = U(n);
        else if (order[i] == 'D') ret[i] = D(n);
    }

    return ret;
}

std::optional<Node> Graph::
L(Node n) {
    if (n.zero_x == 0) return {};
    Node ret(n);

    unsigned char temp = ret.table[ret.zero_y][ret.zero_x - 1];
    ret.table[ret.zero_y][ret.zero_x - 1] = 0;
    ret.table[ret.zero_y][ret.zero_x] = temp;

    ret.zero_x--;
    ret.path += 'L';

    return {ret};
}

std::optional<Node> Graph::
R(Node n) {
    if (n.zero_x == n.size_x - 1) return {};
    Node ret(n);

    unsigned char temp = ret.table[ret.zero_y][ret.zero_x + 1];
    ret.table[ret.zero_y][ret.zero_x + 1] = 0;
    ret.table[ret.zero_y][ret.zero_x] = temp;

    ret.zero_x++;
    ret.path += 'R';

    return {ret};
}

std::optional<Node> Graph::
U(Node n) {
    if (n.zero_y == 0) return {};
    Node ret(n);

    unsigned char temp = ret.table[ret.zero_y - 1][ret.zero_x];
    ret.table[ret.zero_y - 1][ret.zero_x] = 0;
    ret.table[ret.zero_y][ret.zero_x] = temp;

    ret.zero_y--;
    ret.path += 'U';

    return {ret};
}

std::optional<Node> Graph::
D(Node n) {
    if (n.zero_y == n.size_y - 1) return {};
    Node ret(n);

    unsigned char temp = ret.table[ret.zero_y + 1][ret.zero_x];
    ret.table[ret.zero_y + 1][ret.zero_x] = 0;
    ret.table[ret.zero_y][ret.zero_x] = temp;

    ret.zero_y++;
    ret.path += 'D';

    return {ret};
}

File: JAVA/src/main/java/org/example/Graph.java
package org.example;

import java.util.ArrayList;
import java.util.List;

public class Graph {
    private static final String default_order = "LRUD";
    public static Node goal;

    public static boolean isGoal(Node n) {
        return goal.equals(n);
    }

    public static List<Node> neighbors(Node n) {
        return neighbors(n, default_order);
    }

    public static List<Node> neighbors(Node n, String order) {
        List<Node> ret = new ArrayList<>(4);

        for (int i = 0; i < order.length(); i++) {
            Node elem = null;

            if ( order.charAt(i) == 'L' ) elem = L(n);
            else if ( order.charAt(i) == 'R' ) elem = R(n);
            else if ( order.charAt(i) == 'U' ) elem = U(n);
            else if ( order.charAt(i) == 'D' ) elem = D(n);

            if ( elem != null ) ret.add(elem);
        }

        return ret;
    }

    public static Node L(Node n) {
        if ( n.zero_x == 0 ) return null;
        Node ret = new Node(n);

        var temp = ret.table[ ret.zero_y ][ ret.zero_x - 1 ];
        ret.table[ ret.zero_y ][ ret.zero_x - 1 ] = 0;
        ret.table[ ret.zero_y ][ ret.zero_x ] = temp;

        ret.zero_x--;
        ret.path += 'L';

        return ret;
    }

    public static Node R(Node n) {
        if ( n.zero_x == n.size_x - 1 ) return null;
        Node ret = new Node(n);

        var temp = ret.table[ ret.zero_y ][ ret.zero_x + 1 ];
        ret.table[ ret.zero_y ][ ret.zero_x + 1 ] = 0;
        ret.table[ ret.zero_y ][ ret.zero_x ] = temp;

        ret.zero_x++;
        ret.path += 'R';

        return ret;
    }

    public static Node U(Node n) {
        if ( n.zero_y == 0 ) return null;
        Node ret = new Node(n);

        var temp = ret.table[ ret.zero_y - 1 ][ ret.zero_x ];
        ret.table[ ret.zero_y - 1 ][ ret.zero_x ] = 0;
        ret.table[ ret.zero_y ][ ret.zero_x ] = temp;

        ret.zero_y--;
        ret.path += 'U';

        return ret;
    }

    public static Node D(Node n) {
        if ( n.zero_y == n.size_y - 1 ) return null;
        Node ret = new Node(n);

        var temp = ret.table[ ret.zero_y + 1 ][ ret.zero_x ];
        ret.table[ ret.zero_y + 1 ][ ret.zero_x ] = 0;
        ret.table[ ret.zero_y ][ ret.zero_x ] = temp;

        ret.zero_y++;
        ret.path += 'D';

        return ret;
    }

    public static Node reverse(Node n, char operator) {
        if ( operator == 'D' ) return U(n);
        if ( operator == 'U' ) return D(n);
        if ( operator == 'L' ) return R(n);
        if ( operator == 'R' ) return L(n);
        return null;
    }


}
